packages/@jsii/runtime/lib/sync-stdio.ts

```diff
@@ -18,45 +18,41 @@ export class SyncStdio {
   }
 
   public readLine(): string | undefined {
-    if (this.inputQueue.length > 0) {
-      return this.inputQueue.shift();
-    }
-
-    try {
-      const buff = Buffer.alloc(INPUT_BUFFER_SIZE);
-      const read = fs.readSync(STDIN_FD, buff, 0, buff.length, null);
+    const buff = Buffer.alloc(INPUT_BUFFER_SIZE);
+    while (this.inputQueue.length === 0) {
+      try {
+        const read = fs.readSync(STDIN_FD, buff, 0, buff.length, null);
 
-      if (read === 0) {
-        return undefined;
-      }
+        if (read === 0) {
+          return undefined;
+        }
 
-      const str = buff.slice(0, read).toString();
+        const str = buff.slice(0, read).toString();
 
-      for (const ch of str) {
-        if (ch === '\n') {
-          this.inputQueue.push(this.currentLine);
-          this.currentLine = '';
-        } else {
-          this.currentLine += ch;
+        for (const ch of str) {
+          if (ch === '\n') {
+            this.inputQueue.push(this.currentLine);
+            this.currentLine = '';
+          } else {
+            this.currentLine += ch;
+          }
+        }
+      } catch (e) {
+        // HACK: node may set O_NONBLOCK on it's STDIN depending on what kind of input it is made
```

Comment:
We control both sides of the pipe. In what conditions will node use non blocking? Is this something we can configure on the runtime side?

Reply:
According to the linked thread, something as simple as calling `console.log()` may switch `process.stdin` to non-blocking mode.